simulations/common/config.py

```diff
@@ -0,0 +1,176 @@
+import inspect
+from ast import literal_eval as make_tuple
+from collections import namedtuple
+from random import choices
+
+import scipy.stats
+import yaml
+
+PeerType = namedtuple('PeerType', ('config', 'service_map'), defaults=(None, {}))
+
+
+class Dist(object):
+    """Wrapper on scipy.stats functios.
+    Able to generate a value distribution
+    """
+
+    def __init__(self, name: str, params):
+        self.name = name
+        self.params = params
+
+    def to_repr(self):
+        return {self.__class__.__name__: {'name': self.name, 'params': str(self.params)}}
+
+    def __str__(self):
+        return self.__class__.__name__ + ": " + str(self.name) + str(self.params)
+
+    def __repr__(self):
+        return self.__class__.__name__ + ": " + str(self.name) + str(self.params)
+
+    @classmethod
+    def from_repr(cls, yaml_dict):
+        return cls(**yaml_dict)
+
+    def generate(self, n=1):
+        """
+        Generate 'n' random values with given distribution
+        """
+        if self.name == 'sample':
+            weights = self.params['weights'] if 'weights' in self.params else None
+            values = self.params['values'] if 'values' in self.params \
+                else self.params
+            weights = make_tuple(weights) if type(weights) == str else weights
+            values = make_tuple(values) if type(values) == str else values
+            res = choices(values, weights=weights, k=n)
+            return res if n != 1 else res[0]
+
+        dist = getattr(scipy.stats, self.name)
+        param = make_tuple(self.params) if type(self.params) == str else self.params
+        res = dist.rvs(*param[:-2], loc=param[-2], scale=param[-1], size=n)
+        return res if n != 1 else res[0]
+
+    def get(self):
+        return self.generate(1)
+
+
+class Func(object):
+
+    def __init__(self, func):
+        self.func = func
+
+    def to_repr(self):
+        return self.func
+
+    def __str__(self):
+        return repr(self.to_repr())
+
+    def __repr__(self):
+        return repr(self.to_repr())
+
+    @classmethod
+    def from_repr(cls, yaml_dict):
+        return cls(yaml_dict)
+
+    def get(self):
+        return self.func
+
+
+class DistAttr(Dist):
+
+    def get(self):
+        return Dist(self.name, self.params)
+
+
+class Config:
```

Comment:
What's the purpose of this `Config` file? In general, there seems to be a lot of boilerplate code (also in the other classes like `Func` and `Dist`.

Reply:
I will add examples and test to demonstrate it better. 
But the idea is to use python dataclass-style as both configuration file and object with dynamically changing state. 
Every time you execute `Config.get()` it will fetch data from the config declaration and give number. Very useful to simulate changing latency, jitter etc. 

As a downside though, it requires quite some boilerplate, as I also want to be able to write and read from file consistently.